[WIP] Added a variety system for item and block types.

Cybermaxke opened the pull request:
This is a PR as preparation for 1.13, this system allows you to query for specific properties/collections of blocks and items. For example, you can query for all the blocks that are dyed blue. More info will follow.

Related to:
https://github.com/SpongePowered/SpongeAPI/issues/1667
Blocked by: https://github.com/SpongePowered/SpongeAPI/pull/1871

### Features?
- Query ItemTypes, BlockTypes, ItemStacks based on specific varieties/variants.
```java
// Get all the wool item types
VariantCollection<ItemType> woolItems = VariantCollections.ITEMS.query(VarietyQuery.builder()
        .variety(Varieties.WOOL)
        .build());
// OR
woolItems = VariantCollections.ITEMS.query(Varieties.WOOL);

// Get all the wool items that aren't blue or light blue dyed
VariantCollection<ItemType> nonBlueWoolItems = VariantCollections.ITEMS.query(VarietyQuery.builder()
        .variety(Varieties.WOOL)
        .property(new DyeColorProperty(DyeColors.BLUE, Property.Operator.NOTEQUAL))
        .property(new DyeColorProperty(DyeColors.LIGHT_BLUE, Property.Operator.NOTEQUAL))
        .build());
```
- Transform ItemTypes, BlockTypes based on applicable properties.
```java
ItemType someShulkerBox = ItemTypes.WHITE_SHULKER_BOX;
// Want a other color of shulker box, without the need to query?
someShulkerBox = someShulkerBox.transformWith(new DyeColorProperty(DyeColors.BLUE)).get();
```

### Names?
Please give some input on them, we currently have the following possible options:

Varieties:
- VarietyCollection
- VarietyCollections
- Variety
- Varieties
- VarietyQueryable
- VarietyQuery
- VarietyQueryEntry 

OR

Variants:
- VariantCollection
- VariantCollections
- Variant
- Variants
- VariantQueryable
- VariantQuery
- VariantQueryEntry 

OTHER:

Independent:
- MatchOperator

TODO:
- [ ] Add variant matcher

@gabizou

Cybermaxke committed:
Added a variety system for item and block types.

gabizou commented on src/main/java/org/spongepowered/api/variant/VarietyQuery.java:
```diff
@@ -0,0 +1,72 @@
+import java.util.Collection;
+
+// TODO: Better name?
+public interface VarietyQuery {
```
So, if you can put it in your own words, what is this specifically? What separates this from a `Variety`?

gabizou commented on src/main/java/org/spongepowered/api/variant/Variety.java:
```diff
@@ -0,0 +1,31 @@
+
+import org.spongepowered.api.CatalogType;
+
+public interface Variety extends CatalogType {
```
What exactly does this do for us? Is this a "marker for what type of grouping the `VarietyQueryable` is "contained" in? Can we request some properties on a variety? Is it just an assist for making it a translatable object?

gabizou replied:
Could this maybe further define what supported generic types a `VarietyQueryable` can be grouped with this `Variety`? In the sense that `Varieties.BLOCK` would apply to `ItemStack`s?

gabizou commented on src/main/java/org/spongepowered/api/variant/VarietyQuery.java:
```diff
@@ -0,0 +1,72 @@
+     */
+    Collection<Property<?,?>> getProperties();
+
+    interface Builder extends ResettableBuilder<VarietyQuery, Builder> {
```
Is the builder to make a query of multiple varieties? Like are we to use it in this fashion:
```java
Collection<VarietyQuerable> varieties = Sponge.getRegistry().getAll(VarietyQuery.builder()
    .variety(Varieties.BLOCK)
    .variety(Varieties.STAIR)
    .variety(Varieties.SPRUCE)
    .build())
```

Would this make sense?

Cybermaxke replied:
The example above will return all the BlockType's and all the ItemType's with an attached block type. If you want to add a method in the game registry, I would also require a specific return type, so not everything gets (item types, block types, etc.) combined in the returned list.
```java
Collection<ItemType> spruceStairBlockItems = Sponge.getRegistry().getAll(ItemType.class, VarietyQuery.builder()
    .variety(Varieties.BLOCK)
    .variety(Varieties.STAIR)
    .variety(Varieties.SPRUCE)
    .build())
```

gabizou commented on src/main/java/org/spongepowered/api/variant/VarietyQueryable.java:
```diff
@@ -0,0 +1,44 @@
+import org.spongepowered.api.data.property.PropertyHolder;
+
+/**
+ * Represents a {@link CatalogType} that can be filtered in a {@link VariantCollection}.
```
A `CatalogType` specifically? Maybe you mean a `PropertyHolder`?

Cybermaxke replied:
This got leftover when I no longer extended the CatalogType.

gabizou commented on src/main/java/org/spongepowered/api/variant/VariantCollection.java:
```diff
@@ -0,0 +1,107 @@
+ *
+ * @param <T> The variety queryable object type
+ */
+public interface VariantCollection<T extends VarietyQueryable> {
```
IF we're intending to use `VariantCollection` maybe a better name for `Variety` would be `Variant` and then `Varieties` -> `Variants`.

gabizou reviewed (commented):
Some clarifying questions to keep development going.

Cybermaxke committed:
Put some more work in it

- BlockState and ItemStack are queryable
- Added a builder to create VariantCollections
- Added a operator to define how to match the variety
- Added some methods to the game registry

Cybermaxke committed:
Added a TransformablePropertyHolder interface and CookedProperty.

Cybermaxke committed:
transform -> transformWith

ryantheleach commented:
@Cybermaxke I'd like to see some form of serializable variant matcher.

Additionally, does this have anything to do with the 'tags' as added by Mojang in 1.13?

liach commented on src/main/java/org/spongepowered/api/variant/VariantCollections.java:
```diff
@@ -0,0 +1,59 @@
+    /**
+     * A collection with all the available {@link BlockState}s.
+     */
+    public static final VariantCollection<BlockType> BLOCK_STATES = DummyObjectProvider.createFor(VariantCollection.class, "BLOCK_STATES");
```
The generics here is wrong

Cybermaxke commented:
@ryantheleach The tags will be integrated in the implementation. I will add a variant matcher on my todo list.

Cybermaxke commented on src/main/java/org/spongepowered/api/variant/VarietyQuery.java:
```diff
@@ -0,0 +1,181 @@
+ * {@link Variety varieties} and {@link Property properties} on
+ * a {@link VarietyQueryable}.
+ */
+public interface VarietyQuery {
```
Will be renamed to `VarietyMatcher`/`VariantMatcher`

gabizou commented:
> Additionally, does this have anything to do with the 'tags' as added by Mojang in 1.13?


@ryantheleach It's 100% to match exposing the tag system.

ejm commented:
If this is meant to match exposing the tag system, will tags in data packs still be accessible in some form directly?

gabizou commented:
> If this is meant to match exposing the tag system, will tags in data packs still be accessible in some form directly?

Not sure how data packs will be supported on the server, there's a few things currently in development for initial 1.13 development, but this is likely pending some changes from forge to verify client support.

DosMike commented:
I just stumbled over this whole variant issue and thought I might add my 5ct on the latest issue.

How about something similar to Forges OreDict entries. (tho I'm not the biggest fan of OreDict).
Reason being players will always talk about stairs, ingots, blocks, carpet, ...
From my experience with OreDict you'd need 3 properties; Material, Shape and Variant to describe things in Minecraft.
ingotIron, woodStairs, treatedWoosSticks to name examples.

The problem with OreDict serialization is that some mods use camelcase to concatinate words, e.g. an ingot of star metal might become ingotStarMetal.

I would also like to have an option to go from an ItemType to material, shape and/or variant, in case you don't care about stuff like color for an Item.

Cybermaxke commented:
I thought that OreDict was going to be replaced by the Tags system?

DosMike commented:
Oh, I didn't know that. Last time I played with modding in forge was some time ago.

pie-flavor commented:
Will this support the 1.14 addition of entity tags?

JBYoshi commented:
Note: This PR was automatically closed because the `bleeding` branch was deleted. It will need to be reopened to target `api-8` instead.  See https://github.com/SpongePowered/SpongeAPI/pull/1428#issuecomment-539689855.
